File: backend/app/schemas/schedule.py
"""排课 Pydantic Schemas

排课相关 Pydantic 模型
"""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, ConfigDict


class ScheduleBase(BaseModel):
    """排课基础模型"""
    course_id: int = Field(..., description="课程ID")
    teacher_id: int = Field(..., description="教师ID")
    classroom_id: int = Field(..., description="教室ID")
    department_id: Optional[int] = Field(None, description="校区ID")
    start_time: datetime = Field(..., description="开始时间")
    end_time: datetime = Field(..., description="结束时间")
    week_day: Optional[int] = Field(None, ge=1, le=7, description="星期几 1-7")
    recurring_type: Optional[str] = Field(None, max_length=20, description="循环类型: single/weekly/biweekly")
    recurring_id: Optional[str] = Field(None, max_length=50, description="关联的循环ID")
    max_students: int = Field(30, ge=1, description="最大人数")
    enrolled_count: int = Field(0, ge=0, description="已报名人数")
    notes: Optional[str] = Field(None, description="备注")


class ScheduleCreate(ScheduleBase):
    """排课创建模型"""
    pass


class ScheduleUpdate(BaseModel):
    """排课更新模型"""
    course_id: Optional[int] = Field(None, description="课程ID")
    teacher_id: Optional[int] = Field(None, description="教师ID")
    classroom_id: Optional[int] = Field(None, description="教室ID")
    department_id: Optional[int] = Field(None, description="校区ID")
    start_time: Optional[datetime] = Field(None, description="开始时间")
    end_time: Optional[datetime] = Field(None, description="结束时间")
    week_day: Optional[int] = Field(None, ge=1, le=7, description="星期几 1-7")
    recurring_type: Optional[str] = Field(None, max_length=20, description="循环类型: single/weekly/biweekly")
    recurring_id: Optional[str] = Field(None, max_length=50, description="关联的循环ID")
    max_students: Optional[int] = Field(None, ge=1, description="最大人数")
    enrolled_count: Optional[int] = Field(None, ge=0, description="已报名人数")
    status: Optional[int] = Field(None, ge=1, le=4, description="状态: 1:已安排 2:已上课 3:已取消 4:已调课")
    notes: Optional[str] = Field(None, description="备注")


class ScheduleResponse(ScheduleBase):
    """排课响应模型"""
    id: int
    status: int
    created_by: Optional[int]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ScheduleEnroll(BaseModel):
    """学员报名模型"""
    student_id: int = Field(..., description="学员ID")
    count: int = Field(1, ge=1, description="报名人数")


class ScheduleConflictCheck(BaseModel):
    """排课冲突检测模型"""
    teacher_id: int = Field(..., description="教师ID")
    classroom_id: int = Field(..., description="教室ID")
    start_time: datetime = Field(..., description="开始时间")
    end_time: datetime = Field(..., description="结束时间")
    exclude_schedule_id: Optional[int] = Field(None, description="排除的排课ID（用于更新时）")


class ScheduleConflictResponse(BaseModel):
    """排课冲突响应模型"""
    has_conflict: bool = Field(..., description="是否存在冲突")
    teacher_conflicts: list[int] = Field(default_factory=list, description="教师冲突的排课ID列表")
    classroom_conflicts: list[int] = Field(default_factory=list, description="教室冲突的排课ID列表")
    course_conflicts: list[int] = Field(default_factory=list, description="课程冲突的排课ID列表")


__all__ = [
    "ScheduleBase",
    "ScheduleCreate",
    "ScheduleUpdate",
    "ScheduleResponse",
    "ScheduleEnroll",
    "ScheduleConflictCheck",
    "ScheduleConflictResponse",
]

File: miniapp/src/api/schedule.ts
/**
 * 课表相关API
 */
import { request } from './request'

/**
 * 排课基础信息
 */
export interface Schedule {
  id: number
  course_id: number
  course_name?: string
  teacher_id: number
  teacher_name?: string
  classroom_id: number
  classroom_name?: string
  department_id: number | null
  department_name?: string
  start_time: string
  end_time: string
  week_day: number | null
  recurring_type: string | null
  recurring_id: string | null
  max_students: number
  enrolled_count: number
  notes: string | null
  status: number
  created_by: number | null
  created_at: string
}

/**
 * 课表创建参数
 */
export interface ScheduleCreateParams {
  course_id: number
  teacher_id: number
  classroom_id: number
  department_id?: number
  start_time: string
  end_time: string
  week_day?: number
  recurring_type?: string
  max_students?: number
  notes?: string
}

/**
 * 课表更新参数
 */
export interface ScheduleUpdateParams {
  course_id?: number
  teacher_id?: number
  classroom_id?: number
  department_id?: number
  start_time?: string
  end_time?: string
  week_day?: number
  status?: number
  notes?: string
}

/**
 * 课表列表查询参数
 */
export interface ScheduleListParams {
  skip?: number
  limit?: number
  course_id?: number
  teacher_id?: number
  classroom_id?: number
  department_id?: number
  status?: number
  start_date?: string
  end_date?: string
}

/**
 * 课表冲突检测参数
 */
export interface ScheduleConflictCheckParams {
  teacher_id: number
  classroom_id: number
  start_time: string
  end_time: string
  exclude_schedule_id?: number
}

/**
 * 课表冲突检测响应
 */
export interface ScheduleConflictResponse {
  has_conflict: boolean
  teacher_conflicts: number[]
  classroom_conflicts: number[]
  course_conflicts: number[]
}

/**
 * 课表API类
 */
export class ScheduleAPI {
  /**
   * 获取课表列表
   */
  static async list(params?: ScheduleListParams): Promise<Schedule[]> {
    return request.get<Schedule[]>('/schedules', params)
  }

  /**
   * 获取课表详情
   */
  static async getById(scheduleId: number): Promise<Schedule> {
    return request.get<Schedule>(`/schedules/${scheduleId}`)
  }

  /**
   * 创建课表
   */
  static async create(data: ScheduleCreateParams): Promise<Schedule> {
    return request.post<Schedule>('/schedules', data)
  }

  /**
   * 更新课表
   */
  static async update(scheduleId: number, data: ScheduleUpdateParams): Promise<Schedule> {
    return request.put<Schedule>(`/schedules/${scheduleId}`, data)
  }

  /**
   * 删除课表
   */
  static async delete(scheduleId: number): Promise<void> {
    return request.delete<void>(`/schedules/${scheduleId}`)
  }

  /**
   * 检测冲突
   */
  static async checkConflicts(params: ScheduleConflictCheckParams): Promise<ScheduleConflictResponse> {
    return request.post<ScheduleConflictResponse>('/schedules/conflicts/check', params)
  }

  /**
   * 获取今日课表
   */
  static async getTodaySchedules(): Promise<Schedule[]> {
    const today = new Date().toISOString().split('T')[0]
    return request.get<Schedule[]>('/schedules', {
      start_date: today,
      end_date: today
    })
  }

  /**
   * 获取本周课表
   */
  static async getWeekSchedules(weekStart?: string): Promise<Schedule[]> {
    return request.get<Schedule[]>('/schedules', {
      start_date: weekStart
    })
  }

  /**
   * 统计课表数量
   */
  static async count(params?: Partial<ScheduleListParams>): Promise<{ count: number }> {
    return request.get('/schedules/stats/count', params)
  }
}

export default ScheduleAPI
